File: python/andyria/store.py
"""Append-only NDJSON event store for Andyria ledger.

Events are stored as one JSON line per file (events.ndjson).
Event IDs are also written as sentinel files in an index/ directory for O(1) existence checks.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .models import Event


class EventStore:
    """Append-only, content-addressed event ledger backed by NDJSON."""

    def __init__(self, data_dir: Path) -> None:
        """Initialize store at {data_dir}/ledger/."""
        self.ledger_dir = Path(data_dir) / "ledger"
        self.log_path = self.ledger_dir / "events.ndjson"
        self.index_dir = self.ledger_dir / "index"

        self.ledger_dir.mkdir(parents=True, exist_ok=True)
        self.index_dir.mkdir(parents=True, exist_ok=True)

    def append(self, event: Event) -> bool:
        """Append event to log. Returns False if already present (idempotent)."""
        if self.contains(event.id):
            return False

        # Write event as JSON line
        line = event.model_dump_json()
        with open(self.log_path, "a") as f:
            f.write(line + "\n")

        # Write index sentinel
        sentinel = self.index_dir / event.id
        sentinel.write_text(event.id)

        return True

    def contains(self, event_id: str) -> bool:
        """Check if event with this ID is in the log (O(1) via index sentinel)."""
        return (self.index_dir / event_id).exists()

    def load_all(self) -> list[Event]:
        """Load all events from the log in append order."""
        from .models import Event

        if not self.log_path.exists():
            return []

        events = []
        with open(self.log_path) as f:
            for line in f:
                line = line.strip()
                if line:
                    try:
                        event_data = json.loads(line)
                        events.append(Event(**event_data))
                    except (json.JSONDecodeError, TypeError):
                        # Skip malformed lines
                        pass
        return events

    def count(self) -> int:
        """Count of events in the log."""
        if not self.log_path.exists():
            return 0
        return sum(1 for line in open(self.log_path) if line.strip())

File: rust/crates/ledger/src/store.rs
//! Append-only NDJSON event store for the Andyria ledger.
//!
//! Each line of `events.ndjson` is one JSON-serialized `Event`.
//! A flat index directory maps event IDs to a sentinel file for O(1)
//! existence checks without loading the full log.

use std::fs;
use std::io::{BufWriter, Write};
use std::path::{Path, PathBuf};

use thiserror::Error;

use crate::event::Event;

#[derive(Debug, Error)]
pub enum StoreError {
    #[error("io error: {0}")]
    Io(#[from] std::io::Error),
    #[error("serialization error: {0}")]
    Serialization(#[from] serde_json::Error),
    #[error("event not found: {0}")]
    NotFound(String),
}

/// Append-only event log stored as NDJSON.
///
/// Each node maintains its own local log. Peer replication is done by
/// comparing event ID sets and syncing missing events — the content-hash
/// ID ensures idempotent appends across peers.
pub struct EventStore {
    log_path: PathBuf,
    index_dir: PathBuf,
}

impl EventStore {
    pub fn open(data_dir: &Path) -> Result<Self, StoreError> {
        let log_dir = data_dir.join("ledger");
        fs::create_dir_all(&log_dir)?;
        let index_dir = log_dir.join("index");
        fs::create_dir_all(&index_dir)?;

        Ok(Self {
            log_path: log_dir.join("events.ndjson"),
            index_dir,
        })
    }

    /// Append an event to the log. Returns `false` (no-op) for duplicate IDs.
    pub fn append(&self, event: &Event) -> Result<bool, StoreError> {
        if self.contains(&event.id)? {
            return Ok(false);
        }

        let line = serde_json::to_string(event)?;
        let file = fs::OpenOptions::new()
            .create(true)
            .append(true)
            .open(&self.log_path)?;
        let mut writer = BufWriter::new(file);
        writeln!(writer, "{}", line)?;
        writer.flush()?;

        // Write index sentinel
        fs::write(self.index_dir.join(&event.id), event.id.as_bytes())?;
        Ok(true)
    }

    /// Returns `true` if an event with this ID has been appended.
    pub fn contains(&self, event_id: &str) -> Result<bool, StoreError> {
        Ok(self.index_dir.join(event_id).exists())
    }

    /// Load all events in append order.
    pub fn load_all(&self) -> Result<Vec<Event>, StoreError> {
        if !self.log_path.exists() {
            return Ok(Vec::new());
        }
        let content = fs::read_to_string(&self.log_path)?;
        let mut events = Vec::new();
        for line in content.lines() {
            let line = line.trim();
            if !line.is_empty() {
                events.push(serde_json::from_str::<Event>(line)?);
            }
        }
        Ok(events)
    }

    /// Count of events currently in the log.
    pub fn count(&self) -> Result<usize, StoreError> {
        if !self.log_path.exists() {
            return Ok(0);
        }
        let content = fs::read_to_string(&self.log_path)?;
        Ok(content.lines().filter(|l| !l.trim().is_empty()).count())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        crypto::NodeKeyPair,
        event::{EventBuilder, EventType},
    };
    use std::time::{SystemTime, UNIX_EPOCH};

    #[test]
    fn append_and_reload() {
        let dir = tempfile::tempdir().unwrap();
        let store = EventStore::open(dir.path()).unwrap();
        let keypair = NodeKeyPair::generate();
        let ts = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_nanos();

        let event = EventBuilder::new(
            EventType::Checkpoint,
            b"test payload".to_vec(),
            "beacon-001",
            ts,
            "test-node",
        )
        .build(&keypair);

        assert!(store.append(&event).unwrap());
        // Idempotent re-append
        assert!(!store.append(&event).unwrap());

        let loaded = store.load_all().unwrap();
        assert_eq!(loaded.len(), 1);
        assert_eq!(loaded[0].id, event.id);
    }
}
